swap/htlc_test.go

```diff
@@ -306,8 +322,275 @@ func TestHtlcV2(t *testing.T) {
+		valid   bool
+	}{
+		{
+			// Receiver can spend with valid preimage.
+			"success case spend with valid preimage",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = htlc.SuccessSequence()
+				tx.LockTime = uint32(cltvExpiry)
+
+				var trHtlc *HtlcScriptV3
+				trHtlc, ok := htlc.HtlcScript.(*HtlcScriptV3)
+				require.True(t, ok)
+
+				senderPrivKey, _ := btcec.PrivKeyFromBytes(
+					senderPrivKey[:],
+				)
+
+				sig := signTx(
+					tx, senderPrivKey, txscript.NewBaseTapLeaf(trHtlc.ClaimScript),
 				)
+				witness, err := htlc.genSuccessWitness(
+					sig, preimage,
+				)
+				require.NoError(t, err)
+
+				return witness
+			}, true,
+		},
+		{
+			// Receiver can't spend with the valid preimage and with
+			// zero sequence.
+			"success case no spend with valid preimage and zero sequence",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = 0
+
+				var trHtlc *HtlcScriptV3
+				trHtlc, ok := htlc.HtlcScript.(*HtlcScriptV3)
+				require.True(t, ok)
+
+				senderPrivKey, _ := btcec.PrivKeyFromBytes(
+					senderPrivKey[:],
+				)
+
+				sig := signTx(
+					tx, senderPrivKey, txscript.NewBaseTapLeaf(trHtlc.ClaimScript),
+				)
+				witness, err := htlc.genSuccessWitness(
+					sig, preimage,
+				)
+				require.NoError(t, err)
+
+				return witness
+			}, false,
+		},
+		{
+			// Sender can't spend when haven't yet timed out.
+			"timeout case no spend before timeout",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = htlc.SuccessSequence()
+				tx.LockTime = uint32(cltvExpiry) - 1
+
+				var trHtlc *HtlcScriptV3
+				trHtlc, ok := htlc.HtlcScript.(*HtlcScriptV3)
+				require.True(t, ok)
+
+				receiverPrivKey, _ := btcec.PrivKeyFromBytes(
+					receiverPrivKey[:],
+				)
+
+				sig := signTx(
+					tx, receiverPrivKey, txscript.NewBaseTapLeaf(trHtlc.TimeoutScript),
+				)
+
+				witness, err := htlc.GenTimeoutWitness(sig)
+				require.NoError(t, err)
+
+				return witness
+			}, false,
+		},
+		{
+			// Sender can spend after timeout.
+			"timeout case spend after timeout",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = htlc.SuccessSequence()
+				tx.LockTime = uint32(cltvExpiry)
+
+				var trHtlc *HtlcScriptV3
+				trHtlc, ok := htlc.HtlcScript.(*HtlcScriptV3)
+				require.True(t, ok)
+
+				receiverPrivKey, _ := btcec.PrivKeyFromBytes(
+					receiverPrivKey[:],
+				)
+
+				sig := signTx(
+					tx, receiverPrivKey, txscript.NewBaseTapLeaf(trHtlc.TimeoutScript),
+				)
+
+				witness, err := htlc.GenTimeoutWitness(sig)
+				require.NoError(t, err)
+
+				return witness
+			}, true,
+		},
+		{
+			// Receiver can't spend after timeout.
+			"timeout case receiver cannot spend",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = htlc.SuccessSequence()
+				tx.LockTime = uint32(cltvExpiry)
+
+				var trHtlc *HtlcScriptV3
+				trHtlc, ok := htlc.HtlcScript.(*HtlcScriptV3)
+				require.True(t, ok)
+
+				senderPrivKey, _ := btcec.PrivKeyFromBytes(
```

Comment:
nit: the `senderPrivKey` and `receiverPrivKey` could be precalculated outside of the witness function bodies.